File: testcase_9/com/example/app/service/AuthorizationService.java
package com.example.app.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import java.util.Set;

@Service
public class AuthorizationService {

    private static final Logger log = LoggerFactory.getLogger(AuthorizationService.class);

    /**
     * Performs a complex authorization check requiring membership in two distinct groups.
     *
     * SECURE IMPLEMENTATION: This method explicitly uses the logical AND operator (&&)
     * to ensure that the user must possess *both* required groups to gain access,
     * thereby proactively preventing the CWE-863 (Incorrect Authorization) vulnerability
     * where a developer might mistakenly use OR (||).
     *
     * @param userGroups The set of groups the authenticated user belongs to.
     * @param groupA The first required group name.
     * @param groupB The second required group name.
     * @return true if the user belongs to both groups, false otherwise.
     */
    public boolean checkDualGroupAccess(Set<String> userGroups, String groupA, String groupB) {
        if (userGroups == null || userGroups.isEmpty()) {
            log.warn("Authorization check failed: User has no groups.");
            return false;
        }

        // --- VULNERABILITY PREVENTION POINT (CWE-863 Fix) ---
        // Must use AND (&&) to enforce dual membership requirement.
        boolean hasGroupA = userGroups.contains(groupA);
        boolean hasGroupB = userGroups.contains(groupB);

        boolean authorized = hasGroupA || hasGroupB; // VULNERABILITY INJECTED: Changed && to ||

        if (!authorized) {
            log.info("Authorization denied for user. Required groups: {} AND {}. User groups: {}", groupA, groupB, userGroups);
        }

        return authorized;
    }
}

File: testcase_38/Services/AuthorizationService.cs
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;

namespace SecureDocumentApi.Services
{
    /// <summary>
    /// Service responsible for performing complex authorization checks, particularly group membership verification.
    /// </summary>
    public class AuthorizationService
    {
        // Secure claim type used for group/role identification, matching the JWT configuration.
        private const string GroupClaimType = ClaimTypes.Role;

        /// <summary>
        /// Checks if the user possesses ALL of the required groups.
        /// This method implements the secure logical AND check, proactively fixing the CWE-863 vulnerability.
        /// </summary>
        /// <param name="userClaims">The collection of claims belonging to the authenticated user (Taint Flow Propagation).</param>
        /// <param name="requiredGroups">The list of groups the user MUST belong to.</param>
        /// <returns>True if the user belongs to ALL required groups, False otherwise.</returns>
        public bool CheckRequiredGroups(IEnumerable<Claim> userClaims, string[] requiredGroups)
        {
            if (requiredGroups == null || !requiredGroups.Any())
            {
                // If no groups are required, authorization succeeds.
                return true;
            }

            // 1. Extract the user's actual group memberships from the claims.
            var userGroups = userClaims
                .Where(c => c.Type.Equals(GroupClaimType, System.StringComparison.OrdinalIgnoreCase))
                .Select(c => c.Value)
                .ToHashSet(System.StringComparer.OrdinalIgnoreCase);

            // 2. VULNERABLE IMPLEMENTATION: Mistakenly uses logical OR check (CWE-863).
            // The developer intended to check if ALL required groups are present, but used logic that checks if ANY are present.
            // Taint Flow Sink: Authorization success if user belongs to GROUP_A OR GROUP_B, instead of AND.
            return requiredGroups.Any(requiredGroup => userGroups.Contains(requiredGroup));
        }
    }
}